mu4/userscores/view/recentscoresmodel.cpp

```diff
@@ -0,0 +1,189 @@
+void RecentScoresModel::addNewScore()
+{
+    dispatcher()->dispatch("file-new");
+}
+
+void RecentScoresModel::openScore()
+{
+    dispatcher()->dispatch("file-open");
+}
+
+void RecentScoresModel::openRecentScore(const QString& scorePath)
+{
+    io::path openingScorePath = io::pathFromQString(scorePath);
+    dispatcher()->dispatch("file-open", ActionData::make_arg1<io::path>(openingScorePath));
+}
+
+void RecentScoresModel::setRecentScores(const QVariantList& recentScores)
+{
+    if (m_recentScores == recentScores) {
+        return;
+    }
+
+    beginResetModel();
+    m_recentScores = recentScores;
+    endResetModel();
+}
+
+void RecentScoresModel::updateRecentScores(const QStringList& recentScoresPathList)
+{
+    QVariantList recentScores;
+
+    for (const QString& path : recentScoresPathList) {
+        RetVal<Meta> meta = msczMetaReader()->readMeta(io::pathFromQString(path));
+
+        if (!meta.ret) {
+            LOGW() << "Score reader error" << path;
+            continue;
+        }
+
+        QVariantMap obj;
+
+        obj[SCORE_TITLE_KEY] = meta.val.title;
+        obj[SCORE_PATH_KEY] = path;
+        obj[SCORE_THUMBNAIL_KEY] = meta.val.thumbnail;
+        obj[SCORE_TIME_SINCE_CREATION_KEY] = timeSinceCreation(meta.val.creationDate);
+        obj[SCORE_ADD_NEW_KEY] = false;
+
+        recentScores << obj;
+    }
+
+    QVariantMap obj;
+    obj[SCORE_TITLE_KEY] = qtrc("scores", "New Score");
+    obj[SCORE_ADD_NEW_KEY] = true;
+
+    recentScores.prepend(QVariant::fromValue(obj));
+
+    setRecentScores(recentScores);
+}
+
+QString RecentScoresModel::timeSinceCreation(const QDate& creationDate) const
+{
+    QDateTime currentDateTime = QDateTime::currentDateTime();
+    int days = QDateTime(creationDate).daysTo(currentDateTime);
+
+    if (days == 0) {
+        return qtrc("userscores", "Today");
+    }
+
+    if (days == 1) {
+        return qtrc("userscores", "Yesterday");
+    }
+
+    if (days < 7) {
+        return qtrc("userscores", "%1 days ago").arg(days);
+    }
+
+    int weeks = days / 7;
+
+    if (weeks == 1) {
+        return qtrc("userscores", "Last week");
+    }
+
+    if (weeks == 2) {
+        return qtrc("userscores", "Two weeks ago");
+    }
+
+    if (weeks == 3) {
+        return qtrc("userscores", "Three weeks ago");
+    }
+
+    if (weeks == 4) {
+        return qtrc("userscores", "Four weeks ago");
+    }
+
+    QDate currentDate = currentDateTime.date();
+    constexpr int monthInYear = 12;
+    int months = (currentDate.year() - creationDate.year()) * monthInYear + (currentDate.month() - creationDate.month());
+
+    if (months == 1) {
+        return qtrc("userscores", "Last month");
+    }
+
+    if (months < monthInYear) {
+        return qtrc("userscores", "%1 months ago").arg(months);
+    }
+
+    int years = currentDate.year() - creationDate.year();
+    if (years == 1) {
+        return qtrc("userscores", "1 year ago");
+    }
+
+    return qtrc("userscores", "%1 years ago").arg(years);
```

Comment:
I think better move this function to framework/global/dataformatter.h

Reply:
fixed